### DISCUSSION TITLE
misunderstanding, closed.

### DISCUSSION THREAD

**Discussion by mis98zb** [apache/airflow#31627] [Q&A]
I'm trying to find a solution to limit the concurrency for dynamic mapping of task group.
When I read the scheduling codes of airflow 2.6.1, I found a potential bug of max_active_tis_per_dagrun.

In  schedule_job_runner.py, each time doing scheduling, _executable_task_instances_to_queued() calles `__get_concurrency_maps(states=EXECUTION_STATES, session=session)` to initialize concurrency_map with the queued/running TIs from db table task_instance. However, inside the function, it only sets two feilds: dag_active_tasks_map and task_concurrency_map. The feild  task_dagrun_concurrency_map is not set.

The feild  task_dagrun_concurrency_map is only calculated when looping the task_instances_to_examine list, which only contains TIs with state scheduled in db table task_instance.

That is, max_active_tis_per_dagrun can only limit how many scheduled TIs can be run at one time of scheduling, and always ignores the TIs are running.

I'm not sure if I misunsterstood the code, so I just put my question here.

**Comment by pankajastro** [comment]
Look like you are talking about https://github.com/apache/airflow/issues/29084 issue, no?

**Reply by mis98zb** [reply]
Yes. 
The implementation code for issues #29084 may have a defect that it only checks the `max_active_tis_per_dagrun` limitation with scheduled TIs, and ignores the running ones.

**Reply by mis98zb** [reply]
It's my misunderstanding.
Close this.
